tests/cli/shm_ui.c

```diff
@@ -0,0 +1,471 @@
+
+// ******************************** NCURSES ********************************* //
+
+// Initializes ncurses windows
+void init_windows() {
+    // Robot description
+    ROBOT_DESC_WIN = newwin(ROBOT_DESC_HEIGHT, ROBOT_DESC_WIDTH, ROBOT_DESC_START_Y, ROBOT_DESC_START_X);
+    mvwprintw(ROBOT_DESC_WIN, 1, 1, "~~~~~~~~Robot Description~~~~~~~~");
+    // Gamepad
+    GAMEPAD_WIN = newwin(GAMEPAD_HEIGHT, GAMEPAD_WIDTH, GAMEPAD_START_Y, GAMEPAD_START_X);
+    mvwprintw(GAMEPAD_WIN, 1, 1, "~~~~~~~~~~Gamepad State~~~~~~~~~~");
+    // Device info (device data)
+    DEVICE_WIN = newwin(DEVICE_HEIGHT, DEVICE_WIDTH, DEVICE_START_Y, DEVICE_START_X);
+    mvwprintw(DEVICE_WIN, 1, 1, DEVICE_WIN_HEADER);
+
+    refresh();
+}
+
+// Displays the robot description
+void display_robot_desc() {
+    // Gather info from SHM
+    robot_desc_val_t run_mode = robot_desc_read(RUN_MODE);
+    robot_desc_val_t dawn_connection = robot_desc_read(DAWN);
+    robot_desc_val_t shepherd_connection = robot_desc_read(SHEPHERD);
+    robot_desc_val_t gamepad_connection = robot_desc_read(GAMEPAD);
+    robot_desc_val_t start_pos = robot_desc_read(START_POS);
+
+    // Print each field (clear previous value before printing current value)
+    int line = 3;
+    wmove(ROBOT_DESC_WIN, line, 0);
+    wclrtoeol(ROBOT_DESC_WIN);
+    mvwprintw(ROBOT_DESC_WIN, line++, 1, "RUN_MODE = %s", (run_mode == IDLE) ? "IDLE" : (run_mode == AUTO ? "AUTO" : (run_mode == TELEOP ? "TELEOP" : "CHALLENGE")));
+    wclrtoeol(ROBOT_DESC_WIN);
+    mvwprintw(ROBOT_DESC_WIN, line++, 1, "DAWN = %s", (dawn_connection == CONNECTED) ? "CONNECTED" : "DISCONNECTED");
+    wclrtoeol(ROBOT_DESC_WIN);
+    mvwprintw(ROBOT_DESC_WIN, line++, 1, "SHEPHERD = %s", (shepherd_connection == CONNECTED) ? "CONNECTED" : "DISCONNECTED");
+    wclrtoeol(ROBOT_DESC_WIN);
+    mvwprintw(ROBOT_DESC_WIN, line++, 1, "GAMEPAD = %s", (gamepad_connection == CONNECTED) ? "CONNECTED" : "DISCONNECTED");
+    wclrtoeol(ROBOT_DESC_WIN);
+    mvwprintw(ROBOT_DESC_WIN, line++, 1, "START_POS = %s", (start_pos == LEFT) ? "LEFT" : "RIGHT");
+
+    // Box and refresh
+    box(ROBOT_DESC_WIN, 0, 0);
+    wrefresh(ROBOT_DESC_WIN);
+}
+
+/**
+ * Reads the current gamepad state and displays it.
+ * Arguments:
+ *    joystick_names: the array of the joystick names
+ *    button_names: the array of button names
+ */
+void display_gamepad_state(char** joystick_names, char** button_names) {
+    // Start at line after the header
+    int line = 2; // Note that this is just a row number; Must use wmove() to actually move the cursor to this line
+    wmove(GAMEPAD_WIN, line, 0);
+    wclrtoeol(GAMEPAD_WIN); // Clear "No gamepad connected"
+
+    // Read gamepad state if gamepad is connected
+    uint32_t pressed_buttons = 0;
+    float joystick_vals[4] = {0};
+    int gamepad_connected = (robot_desc_read(GAMEPAD) == CONNECTED) ? 1 : 0;
+    if (gamepad_connected) {
```

Comment:
No need to have this int `gamepad_connected` and the ternary expression. Just directly compare the robot desc to `CONNECTED` in the if statement.

Reply:
Makes sense. I had the extra variable because I used it later in this function multiple times, but I removed them.